peps/pep-0787.rst

```diff
@@ -0,0 +1,279 @@
+
+.. code-block:: python
+
+    # Unsafe with f-strings:
+    os.system(f"echo {message_from_user}")
+
+    # Safe with t-strings:
+    subprocess.run(t"echo {message_from_user}", shell=True)
+
+Currently, developers must choose between convenience (using f-strings with potential
+security risks) and safety (using more verbose, list-based APIs). By adding native t-string
+support to the :mod:`subprocess` module, we provide a practical reference implementation that
+demonstrates the value of t-strings while addressing a common security concern.
+
+Rationale
+=========
+
+The subprocess module is an ideal candidate for t-string support for several reasons:
+
+1. Command injection vulnerabilities in shell commands are a well-known security risk.
+2. The subprocess module already supports both string and list-based command specifications.
+3. There's a natural mapping between t-strings and proper shell escaping that provides both convenience and safety.
+4. It serves as a practical showcase for t-strings that developers can immediately understand and appreciate.
+
+By extending subprocess to handle t-strings natively, we make it easier to write secure code without sacrificing
+the convenience that led many developers to use potentially unsafe f-strings.
+
+Specification
+=============
+
+This PEP proposes two main additions to the standard library:
+
+1. A new ``sh()`` renderer function in the :mod:`shlex` module for safe shell command construction
+2. Adding t-string support to the :mod:`subprocess` module's core functions,
+    particularly :class:`subprocess.Popen`, :func:`subprocess.run`, and other related functions
+    that accept a command argument
+
+
+Renderer for shell escaping added to :mod:`shlex`
+-------------------------------------------------
+
+As a reference implementation, a renderer for safe POSIX shell escaping will be added to
+the :mod:`shlex` module. This renderer would be called ``sh`` and would be equivalent to
+calling ``shlex.quote`` on each field value in the template literal.
+
+Thus:
+
+.. code-block:: python
+
+  os.system(shlex.sh(t"cat {myfile}"))
+
+would have the same behavior as:
+
+.. code-block:: python
+
+  os.system("cat " + shlex.quote(myfile)))
+
+The implementation would be:
+
+.. code-block:: python
+
+  def sh(template: TemplateLiteral):
+      def render_field(value, format_spec, conversion_spec)
+          field_text = format(value, format_spec, conversion_spec)
+          return quote(field_text)
+      return template.render(render_field=render_field)
+
+The addition of ``shlex.sh`` will NOT change the existing admonishments in the
+:mod:`subprocess` documentation that passing ``shell=True`` is best avoided, nor the
+reference from the :func:`os.system` documentation to the higher level ``subprocess`` APIs.
+
+Then implementation would look like:
+
+.. code-block:: python
+
+    from string.templatelib import Template, Interpolation
+
+    def sh(template: Template) -> str:
```

Comment:
I don't understand why there are two different implementations here, both of which are apparently supposed to become `shlex.sh`.

Reply:
Copy-and-paste error from PEP 501 (the first implementation uses the PEP 501 implementation, not the PEP 750 one)